### PR TITLE
[llvm] Use LlvmProgramImpl::cache_data_ to store compiled kernel info

### PR DESCRIPTION
**Pull request by jim19930609**
Related issue = #4800

LLVM AOT and Compute Graph captures compiled information for serialization and deserialization. 

In a previous PR, we used LlvmProgramImpl::cache_data_ to save the compiled Fields ([PR#5120](https://github.com/taichi-dev/taichi/pull/5120)). In this PR, we further made LlvmProgramImpl::cache_data_ to save compiled Kernel Info for later use in AOT and Compute Graph.

### CODE REVIEW

File: taichi/runtime/program_impls/llvm/llvm_program.h [hunk 1]
```diff
@@ -223,10 +223,51 @@ class LlvmProgramImpl : public ProgramImpl {
...
+
+    // 2. Destructs runtime_exec_
+    runtime_exec_.reset();
+  }
+
  private:
   std::size_t num_snode_trees_processed_{0};
-  LlvmOfflineCache cache_data_;
   std::unique_ptr<LlvmRuntimeExecutor> runtime_exec_;
+  LlvmOfflineCache cache_data_;
```
Comment (ailzhang) [inline taichi/runtime/program_impls/llvm/llvm_program.h:270]: nit: maybe use a unique_ptr for this?

**Review by ailzhang — APPROVED** [review]
LGTM!

**Review by lin-hitonami — APPROVED** [review]
LGTM!

File: taichi/runtime/program_impls/llvm/llvm_program.h (hunk 1 above)
Comment (jim19930609) [reply taichi/runtime/program_impls/llvm/llvm_program.h:270]: yup, let me put this in a separate PR

**Pull request merged**